File: app/src/main/java/com/ladsoft/bakingapp/di/module/RepositoryModule.java
package com.ladsoft.bakingapp.di.module;

import com.ladsoft.bakingapp.data.database.AppDatabase;
import com.ladsoft.bakingapp.data.database.repository.DatabaseIngredientRepository;
import com.ladsoft.bakingapp.data.database.repository.DatabaseRecipeRepository;
import com.ladsoft.bakingapp.data.database.repository.DatabaseStepRepository;
import com.ladsoft.bakingapp.data.network.api.BakingAppApiModule;
import com.ladsoft.bakingapp.data.repository.RecipeRepository;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public abstract class RepositoryModule {
    @Provides @Singleton
    static RecipeRepository providesRecipeRepository(BakingAppApiModule apiModule) { return new RecipeRepository(apiModule); }

    @Provides @Singleton
    static DatabaseRecipeRepository providesLocalRecipeRepository(AppDatabase appDatabase) { return new DatabaseRecipeRepository(appDatabase); }

    @Provides @Singleton
    static DatabaseIngredientRepository providesLocalIngredientRepository(AppDatabase appDatabase) { return new DatabaseIngredientRepository(appDatabase); }

    @Provides @Singleton
    static DatabaseStepRepository providesLocalStepRepository(AppDatabase appDatabase) { return new DatabaseStepRepository(appDatabase); }
}

File: app/src/main/java/com/ladsoft/bakingapp/di/component/RepositoryModule.kt
package com.ladsoft.bakingapp.di.component

import com.ladsoft.bakingapp.data.database.AppDatabase
import com.ladsoft.bakingapp.data.database.repository.DatabaseIngredientRepository
import com.ladsoft.bakingapp.data.database.repository.DatabaseRecipeRepository
import com.ladsoft.bakingapp.data.database.repository.DatabaseStepRepository
import com.ladsoft.bakingapp.data.repository.RecipeRepository
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class RepositoryModule {
    @Provides @Singleton
    fun providesRecipeRepository(): RecipeRepository = RecipeRepository()

    @Provides @Singleton
    fun providesLocalRecipeRepository(appDatabase: AppDatabase): DatabaseRecipeRepository = DatabaseRecipeRepository(appDatabase)

    @Provides @Singleton
    fun providesLocalIngredientRepository(appDatabase: AppDatabase): DatabaseIngredientRepository = DatabaseIngredientRepository(appDatabase)

    @Provides @Singleton
    fun providesLocalStepRepository(appDatabase: AppDatabase): DatabaseStepRepository= DatabaseStepRepository(appDatabase)
}